### PR TITLE
[Serve] Stop the http proxy actor after it's drained

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
- Stop the http proxy actor after it's drained so the node can be idle terminated under the new idle definition (zero-cpu actor is not idle).
- Refactor the http proxy state manager (to be more similar to the deployment state manager)
    * Change http proxy actor max_restarts to 0 to let the state manager manage the lifecycle of the actor.
    * Set the draining flag of the proxy via the manager instead of LongPollClient so that the draining states in the manager and in the actor are in sync.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/client/src/type/serve.ts
Comment (edoakes) [inline dashboard/client/src/type/serve.ts:95]: Do we need to expose `DRAINED`? This will only be a momentary state just before the proxy is shut down.

File: python/ray/serve/_private/constants.py
Comment (edoakes) [inline python/ray/serve/_private/constants.py:124]: nit: 
```suggestion
PROXY_MIN_DRAINING_PERIOD_S = (
```
verbally you would said "the minimum drain period"

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:299]: maybe better to use `None` instead of magic constant of `0`

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:302]:
```suggestion
        self._min_drain_period_timer = None
```

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:319]: I think the draining timer is completely unnecessary -- can't you just check if `time.time() - self._drain_start_time > MIN_DRAIN_PERIOD`?

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:993]: please update docstring re: draining behavior

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:999]: this still looks pretty hacky to me... IMO the more robust solution would be to return a status code from this health check, then have the controller make the decision to remove the proxy

I won't block the PR on this change if it's much harder to model it that way, but having an implicit signal of the actor killing itself doesn't sit very well with me

File: python/ray/serve/controller.py
Comment (edoakes) [inline python/ray/serve/controller.py:348]: why is this block moved?

File: python/ray/serve/controller.py
Comment (edoakes) [inline python/ray/serve/controller.py:337]: why rename from active_nodes -> http_proxy_nodes? This is a little bit of an abstraction leak

File: dashboard/client/src/type/serve.ts
Comment (jjyao) [reply dashboard/client/src/type/serve.ts:95]: We don't. I added it purely for consistency to keep them in sync. But I guess I can just add a comment and remove it from frontend.

File: python/ray/serve/controller.py
Comment (jjyao) [reply python/ray/serve/controller.py:337]: I want to make it explicit that this one is for deciding the nodes where proxies should run: this is based on active nodes returned from deployment state manager and other factors (draining nodes).

File: python/ray/serve/controller.py
Comment (jjyao) [reply python/ray/serve/controller.py:348]: Dependency wise, http proxy state manager depends on the latest active nodes from deployment state manager so I think it's better to order the code following the dependency.

Comment (edoakes) [reply python/ray/serve/controller.py:348]: with the new drain protocol, we shouldn't need to infer the draining nodes from deployment state manager anymore right? won't the ray scheduler/autoscaler decide for us?

File: python/ray/serve/controller.py
Comment (edoakes) [reply python/ray/serve/controller.py:337]: got it, makes sense

File: python/ray/serve/_private/http_proxy.py
Comment (jjyao) [reply python/ray/serve/_private/http_proxy.py:319]: oops, yea, its no longer needed.

File: python/ray/serve/controller.py
Comment (jjyao) [reply python/ray/serve/controller.py:348]: For draining nodes, you don't get it from deployment state manager but we still get active nodes from it. http proxy nodes is basically active_nodes - draining_nodes.

**Review by GeneDer — COMMENTED** [review]
For my understanding, this change will just not start a proxy on inactive node? Also, was this manually tested for the new beheavior?

File: python/ray/serve/_private/http_proxy.py
Comment (jjyao) [reply python/ray/serve/_private/http_proxy.py:294]: If there is no replicas, the actor will be removed so the object will be released automatically.

**Comment by jjyao** [comment]
> For my understanding, this change will just not start a proxy on inactive node?

Yup, that's the main change. The reason is that in the future, a node running zero cpu actor will no longer be considered idle so we need to fully stop it.

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:316]: pls add docstring describing semantics

File: python/ray/serve/_private/http_proxy.py
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:884]: proper punctuation & grammar please
```suggestion
            f"Proxy actor {ray.get_runtime_context().get_actor_id()} "
            f"starting on on node {node_id}."
```

**Review by edoakes — COMMENTED** [review]
Core code changes LGTM, some pending comments about style/docstrings.

Unless I'm missing something, it looks like we're missing some integration test cases that verify the proxy is stopped/started as expected. Should have tests for the "happy paths" (replica placed on node -> proxy started & replica removed from node -> proxy stopped) as well as oscillating back and forth as we manually tested yesterday.

File: python/ray/serve/_private/http_state.py
Comment (GeneDer) [inline python/ray/serve/_private/http_state.py:187]: This is super helpfully, thanks for adding!

**Review by GeneDer — COMMENTED** [review]
Ditto to Ed's comment, just make sure to address docstring and testing. Everything else LGTM.

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/tests/test_standalone_3.py
Comment (GeneDer) [inline python/ray/serve/tests/test_standalone_3.py:436]: Should we check one being drained? 🙃

**Review by GeneDer — APPROVED** [review]

File: python/ray/serve/tests/test_standalone_3.py
Comment (jjyao) [reply python/ray/serve/tests/test_standalone_3.py:435]: Currently it's 10s. Based on my experimental, it's the shortest time to be not flaky.

File: python/ray/serve/tests/test_standalone_3.py
Comment (jjyao) [reply python/ray/serve/tests/test_standalone_3.py:436]: It's checked. DRAINED means the proxy is removed so here we have total 2 healthy proxies.

**Pull request merged**
